RecursiveExtractor.Tests/ExtractorTests.cs

```diff
@@ -41,6 +42,64 @@ public void ExtractArchive(string fileName, int expectedNumFiles = 26)
             Assert.IsTrue(results.Count() == expectedNumFiles);
         }
 
+        [DataTestMethod]
+        [DataRow("SharedEncrypted.zip")]
+        [DataRow("SharedEncrypted.7z")]
+        [DataRow("SharedEncrypted.rar4")]
+        // [DataRow("SharedEncrypted.rar")] // RAR5 is not yet supported by SharpCompress: https://github.com/adamhathcock/sharpcompress/issues/517
+        public void ExtractEncryptedArchive(string fileName, int expectedNumFiles = 26)
+        {
+            var extractor = new Extractor();
+            var path = Path.Combine(Directory.GetCurrentDirectory(), "TestData", fileName);
+            var results = extractor.ExtractFile(path, new ExtractorOptions()
+            {
+                Passwords = new Dictionary<Regex, List<string>>()
+                {
+                    {
+                        new Regex(".*"), 
+                        new List<string>()
+                        {
+                            "AnIncorrectPassword",
+                            "TheMagicWordIsPotato"
+                        } 
+                    } 
+                }
+            }).ToList();
+            Assert.IsTrue(results.Count() == expectedNumFiles);
+        }
+
+        [DataTestMethod]
+        [DataRow("SharedEncrypted.zip")]
+        [DataRow("SharedEncrypted.7z")]
+        [DataRow("SharedEncrypted.rar4")]
+        // [DataRow("SharedEncrypted.rar")] // RAR5 is not yet supported by SharpCompress: https://github.com/adamhathcock/sharpcompress/issues/517
+
+        public async Task ExtractEncryptedArchiveAsync(string fileName, int expectedNumFiles = 26)
+        {
+            var extractor = new Extractor();
+            var path = Path.Combine(Directory.GetCurrentDirectory(), "TestData", fileName);
+            var results = extractor.ExtractFileAsync(path, new ExtractorOptions()
+            {
+                Passwords = new Dictionary<Regex, List<string>>()
+                {
+                    {
+                        new Regex(".*"),
```

Comment:
Unclear why not include a password for testing all three options mentioned in the documentation i.e.      { new Regex("\.zip"), new List<string>(){ "PasswordForZipFiles" } },
            { new Regex("\.7z"), new List<string>(){ "PasswordFor7zFiles" } },
            { new Regex(".*"), new List<string>(){ "PasswordForAllFiles" } }
 and why regex for the pattern over a straight string?

Reply:
The pattern won't always be a file extension that's just the example I provided.  The pattern could be whatever you want.  If you have a different password for files named "Archive_somepattern.zip" and "Archive_someotherpattern.zip" you can specify both.

Reply:
The test now uses more than 1 pattern a different password for the 7z arhive.

Reply:
Here's an example of when you'd use regexes.

I have a .zip that contains a number of encrypted .zip files.  Each zip file is named with a timestamp for when it was created.  For the first half of 2019 I used the password "Potato" and the second half I used "Tomato". But sometime in 2020 I switched to "Celery".

My zips are named like "MMDDYYYY.zip".

```
{ new Regex("^0[1-6][0-3][0-9]2019.zip"), new List(){ "Potato" } },
{ new Regex("^((1[012])|(0[789]))[0-3][0-9]2019.zip"), new List(){ "Tomato" } },
{ new Regex("^[01][0-9][0-3][0-9]2020.zip"), new List(){ "Celery", "Tomato" } },
```

Reply:
The main downside to additionally providing incorrect passwords as well as correct ones is wasted computation time and raising more exceptions (also computation time).  

So in theory we could do this with a Contains on `.zip` but that would mean in this case testing passwords we now won't work against some archives.